src/pages/invoices/common/hooks/useResolveInputField.tsx

```diff
@@ -45,18 +49,82 @@ interface Props {
   resource: ProductTableResource;
   type: 'product' | 'task';
   relationType: RelationType;
-  onLineItemChange: (index: number, lineItem: InvoiceItem) => unknown;
+  onLineItemChange: (
+    index: number,
+    lineItem: InvoiceItem
+  ) => InvoiceItem[] | Promise<InvoiceItem[]>;
   onLineItemPropertyChange: (
     key: keyof InvoiceItem,
     value: unknown,
     index: number
-  ) => unknown;
+  ) => InvoiceItem[] | Promise<InvoiceItem[]>;
+  createItem: () => unknown;
+  deleteLineItem: (index: number) => unknown;
 }
 
+export const isLineItemEmpty = (lineItem: InvoiceItem) => {
+  if (
+    !lineItem.cost &&
+    !lineItem.quantity &&
+    !lineItem.notes &&
+    !lineItem.product_key
+  ) {
+    return true;
+  }
+
+  return false;
+};
+
 export function useResolveInputField(props: Props) {
   const [inputCurrencySeparators, setInputCurrencySeparators] =
     useState<DecimalInputSeparators>();
 
+  const isAnyExceptLastLineItemEmpty = (items: InvoiceItem[]) => {
+    const filteredItems = items.filter(
+      (item, index) => index !== items.length - 1
+    );
+
+    return filteredItems.some((lineItem) => isLineItemEmpty(lineItem));
+  };
+
+  const cleanLineItemsList = (lineItems: InvoiceItem[]) => {
+    let typeId = InvoiceItemType.Product;
+
+    if (props.type === 'task') {
+      typeId = InvoiceItemType.Task;
+    }
+
+    const typeFilteredLineItems = lineItems.filter(
+      ({ type_id }) => type_id === typeId
+    );
+
+    const lineItemsLength = typeFilteredLineItems.length;
+
+    const lastLineItem = typeFilteredLineItems[lineItemsLength - 1];
+
+    if (lineItemsLength > 0) {
+      if (
+        !isAnyExceptLastLineItemEmpty(typeFilteredLineItems) &&
+        !isLineItemEmpty(lastLineItem)
+      ) {
+        props.createItem();
+      }
+
+      if (
+        isAnyExceptLastLineItemEmpty(typeFilteredLineItems) &&
+        isLineItemEmpty(lastLineItem)
+      ) {
+        const lastLineItemIndex = lineItems.indexOf(
+          typeFilteredLineItems[lineItemsLength - 1]
+        );
+
+        if (lastLineItemIndex > -1) {
+          props.deleteLineItem(lastLineItemIndex);
+        }
+      }
+    }
+  };
```

Comment:
This seems like a fairly expensive operation. Is it possible to memoize this with `useCallback`?

Reply:
Sure, the logic has been changed.